File: app/src/main/java/com/scatl/uestcbbs/module/setting/view/OpenSourceActivity.java
package com.scatl.uestcbbs.module.setting.view;

import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.RecyclerView;

import android.view.View;
import android.view.animation.AnimationUtils;
import android.view.animation.LayoutAnimationController;
import android.widget.ImageView;

import com.alibaba.fastjson.JSONObject;
import com.scatl.uestcbbs.R;
import com.scatl.uestcbbs.base.BaseActivity;
import com.scatl.uestcbbs.base.BasePresenter;
import com.scatl.uestcbbs.custom.MyLinearLayoutManger;
import com.scatl.uestcbbs.entity.OpenSourceBean;
import com.scatl.uestcbbs.module.setting.adapter.OpenSourceAdapter;
import com.scatl.uestcbbs.util.CommonUtil;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;

public class OpenSourceActivity extends BaseActivity {

    private Toolbar toolbar;
    private RecyclerView recyclerView;
    private OpenSourceAdapter openSourceAdapter;

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_open_source;
    }

    @Override
    protected void findView() {
        toolbar = findViewById(R.id.open_source_toolbar);
        recyclerView = findViewById(R.id.open_source_rv);

    }

    @Override
    protected void initView() {

        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        openSourceAdapter = new OpenSourceAdapter( R.layout.item_open_source);
        openSourceAdapter.setHasStableIds(true);
        recyclerView.setLayoutManager(new MyLinearLayoutManger(this));
        recyclerView.setAdapter(openSourceAdapter);
        LayoutAnimationController layoutAnimationController = AnimationUtils.loadLayoutAnimation(this, R.anim.layout_animation_from_top);
        recyclerView.setLayoutAnimation(layoutAnimationController);

        setData();
    }

    @Override
    protected BasePresenter initPresenter() {
        return null;
    }

    @Override
    protected void setOnItemClickListener() {
        openSourceAdapter.setOnItemClickListener((adapter, view, position) -> {
            if (view.getId() == R.id.item_open_source_cardview) {
                CommonUtil.openBrowser(OpenSourceActivity.this, openSourceAdapter.getData().get(position).link);
            }
        });
    }

    private void setData() {
        String data = "[]";
        InputStream is;
        try {
            is = getAssets().open("open_source_projects.json");
            ByteArrayOutputStream baos = new ByteArrayOutputStream();

            int len;
            byte[] buffer = new byte[1024];
            while ((len = is.read(buffer)) != -1) {
                baos.write(buffer, 0, len);
            }
            data = baos.toString();
            is.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

        List<OpenSourceBean> openSourceBeanList = JSONObject.parseArray(data, OpenSourceBean.class);
        openSourceAdapter.addData(openSourceBeanList);


    }
}

File: app/src/main/java/com/scatl/uestcbbs/module/setting/view/OpenSourceActivity.kt
package com.scatl.uestcbbs.module.setting.view

import android.view.View
import android.view.animation.AnimationUtils
import com.chad.library.adapter.base.BaseQuickAdapter
import com.scatl.uestcbbs.R
import com.scatl.uestcbbs.base.BaseVBActivity
import com.scatl.uestcbbs.databinding.ActivityOpenSourceBinding
import com.scatl.uestcbbs.entity.OpenSourceBean
import com.scatl.uestcbbs.module.setting.adapter.OpenSourceAdapter
import com.scatl.uestcbbs.module.setting.presenter.OpenSourcePresenter
import com.scatl.uestcbbs.util.CommonUtil
import com.scwang.smart.refresh.layout.api.RefreshLayout

/**
 * Created by sca_tl at 2023/6/6 16:17
 */
class OpenSourceActivity: BaseVBActivity<OpenSourcePresenter, OpenSourceView, ActivityOpenSourceBinding>(), OpenSourceView {

    private lateinit var openSourceAdapter: OpenSourceAdapter

    override fun getViewBinding() = ActivityOpenSourceBinding.inflate(layoutInflater)

    override fun initPresenter() = OpenSourcePresenter()

    override fun initView(theftProof: Boolean) {
        super.initView(true)
        openSourceAdapter = OpenSourceAdapter(R.layout.item_open_source)
        mBinding.recyclerView.layoutAnimation = AnimationUtils.loadLayoutAnimation(this, R.anim.layout_animation_scale_in)
        mBinding.recyclerView.adapter = openSourceAdapter
        mBinding.statusView.loading()
        mBinding.refreshLayout.setEnableLoadMore(false)
        mPresenter?.getOpenSourceData()
    }

    override fun setOnItemClickListener() {
        openSourceAdapter.setOnItemClickListener { adapter, view, position ->
            CommonUtil.openBrowser(getContext(), openSourceAdapter.data[position].link)
        }
    }

    override fun onRefresh(refreshLayout: RefreshLayout) {
        mPresenter?.getOpenSourceData()
    }

    override fun onGetOpenSourceDataSuccess(data: List<OpenSourceBean>) {
        mBinding.statusView.success()
        mBinding.refreshLayout.finishRefresh()
        openSourceAdapter.setNewData(data)
        mBinding.recyclerView.scheduleLayoutAnimation()
    }

    override fun onGetOpenSourceDataError(msg: String?) {
        mBinding.statusView.error(msg)
    }

    override fun getContext() = this
}